Source/Randomizer.cpp

```diff
@@ -132,16 +132,50 @@ void Randomizer::Randomize() {
     RandomizeChallenge();
 
     // Content swaps -- must happen before squarePanels
-    Randomize(upDownPanelsSetZero, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetOne, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetTwo, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetThree, SWAP::LINES | SWAP::COLORS);
-    Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
-    Randomize(leftForwardRightPanelsSetOne, SWAP::LINES | SWAP::COLORS);
-    Randomize(leftForwardRightPanelsSetTwo, SWAP::LINES | SWAP::COLORS);
-
-    Randomize(quarryLaserOptions, SWAP::LINES | SWAP::COLORS);
-    Randomize(squarePanels, SWAP::LINES | SWAP::COLORS);
+    if (_doubleRandomizer) {
+        // The pool that Tutorial Back Left is in has some panels that may not
+        // be solveable in the down position after using Sigma's randomizer. We
+        // do not want to swap any such panel with Tutorial Back Left, because
+        // that would make it impossible to exit Tutorial.
+        Randomize(upDownPanelsSetThreeDoubleMode, SWAP::LINES | SWAP::COLORS);
+        // The four pivot panels in Treehouse must be solveable in the up, left,
+        // and right positions. However, the other panels in the pools those
+        // panels are found in may not be solveable in all three directions
+        // after using Sigma's randomizer. For safety, we will only swap the
+        // pivot panels amongst themselves.
+        Randomize(treehousePivotSet, SWAP::LINES | SWAP::COLORS);
+        // This will additionally shuffle the UTM perspective puzzles amongst
+        // themselves.
+        Randomize(utmPerspectiveSet, SWAP::LINES | SWAP::COLORS);
+        // In order to prevent a situation where access to a Symmetry Laser
+        // Yellow is blocked by its corresponding Blue panel, we will only
+        // shuffle these six panels amongst themselves.
+        Randomize(symmetryLaserYellows, SWAP::LINES | SWAP::COLORS);
+        Randomize(symmetryLaserBlues, SWAP::LINES | SWAP::COLORS);
+        // Prevent the fourth UTM perspective puzzle from being shuffled, as it
+        // is unknown whether any of the other puzzles in these pools can be
+        // solved there after using Sigma's randomizer.
+        Randomize(upDownPanelsSetZeroDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetOneDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetTwoDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
+        // Many puzzles either crash the game or do not solve properly when
+        // swapped with Swamp Entry. To make things simpler, we will just remove
+        // that panel from both pools it is found in.
+        Randomize(quarryLaserOptionsDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(squarePanelsDoubleMode, SWAP::LINES | SWAP::COLORS);
+    }
+    else {
```

Comment:
```suggestion
    } else {
```
[nit] style in this codebase